### PR TITLE
Remove C#/VB dependency on inline hints.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is necessary if F# or TS want to add thse sorts of hints to their experience in VS.  Note: This is a followup to my previous PRs.  It shoudl not be reviewed until that goes in.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @cartermp @DanielRosenwasser  THis allows experiences like holding down ctrl-alt to see tags like this:

image

Hints can be shown in-situ with text (like `v2: `) or can replace text entirely (like `List<int>`).  Can be good for allowing people to quickly see this information in a lightweight fashion.  Once you release ctrl-alt, they go away.  

You can also just have permanently showing hints.  Though we advise that users be able to opt-into/out of that in case they don't like it.

**Comment by cartermp** [comment]
Tagging myself here to add this API to the F# shim. This is neat!

**Comment by cartermp** [comment]
FWIW this is a pretty damn useful feature for F#, we may want to experiment with having it on by default

**Comment by DanielRosenwasser** [comment]
> FWIW this is a pretty damn useful feature for F#, we may want to experiment with having it on by default

TypeScript/JavaScript would have to be careful - semantic highlighting done naively tended to have issues since we only have one fully semantic thread. The key combo is probably more desirable.

(CC @mjbvz @ryancavanaugh @minestarks @uniqueiniquity)

**Comment by CyrusNajmabadi** [comment]
> semantic highlighting done naively tended to have issues since we only have one fully semantic thread

:'(

I recommend having interruptible semantic operations that you can yield back on but then resume later.

**Comment by DanielRosenwasser** [comment]
Sorry, our entire compiler/LS is [blue](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).

**Comment by CyrusNajmabadi** [comment]
> Sorry, our entire compiler/LS is blue.

THe compiler can stay blue.  It would require making the LS async/reenterable.  However, i do not see that being that difficult to do.

**Comment by minestarks** [comment]
Is there a corresponding LSP spec for this?

**Comment by CyrusNajmabadi** [comment]
> Is there a corresponding LSP spec for this?

Not yet.  Primarily because i don't want to preemptively write sucha  spec, and then have to change it as we learn about new requirements.  For example, when i started, the feature was just about inserting some text at a particular location in the document.  But it has already morphed three times:

1. it can now *replace* a span of text.  We use this to replace `var` with the actual type while the user is holding Ctrl-Alt.
2. it can produce classified text (as you can see in teh picture)
3. you can hover over the tag to get rich documentation.  i.e.:

image

This is why i'd liek to see how F#/TS/JS want to use this.  If you run into changes you'd like we can design through all of that.  once we feel like we're in a good place, we coudl then bring this to be an LSP addition :)

**Review by cartermp — APPROVED** [review]
Actually if it's not too big a deal @CyrusNajmabadi - if you wanted to just create F# shims for the relevant types and interfaces then that would be great. I imagine this would be the best subfolder: https://github.com/dotnet/roslyn/tree/master/src/Tools/ExternalAccess/FSharp/Editor

**Comment by CyrusNajmabadi** [comment]
> Actually if it's not too big a deal @CyrusNajmabadi - if you wanted to just create F# shims for the relevant types and interfaces then that would be great.

Will do once this PR goes in.

File: src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTag.cs [hunk 1]
```diff
@@ -71,76 +72,38 @@ internal class InlineHintsTag : IntraTextAdornmentTag
...
+            var document = _span.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document != null)
             {
-                var document = _span.Snapshot.TextBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
-
-                if (document != null)
+                var taggedText = await _hint.GetDescriptionAsync(document, cancellationToken).ConfigureAwait(false);
+                if (!taggedText.IsDefaultOrEmpty)
                 {
-                    var compilation = await document.Project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core.Wpf/InlineHints/InlineHintsTag.cs:98]: all this logic moved in the C#/VB impl of GetDescriptionAsync

**Review by davidwengier — APPROVED** [review]

**Pull request merged**